File: 3°Semestre/TecnicasProgram-II/Exemplos/Singleton/Singleton02.py
class Aluno:
# _XXXXXXXX -> Indica que o atributo é protegido
    _instancia = None
    _inicializado = False

# '*' -> Desempacota sequencias (lista, tuplas, etc)
# '**' -> Desempacota dicionários (estruturas)
    def __new__(cls, *args, **kwargs):
        if not cls._instancia:
            cls._instancia = super().__new__(cls, *args, **kwargs)
        return cls._instancia
    
    def __init__(self):
        if not self._inicializado:
            self._inicializado = True
            self.nome = ""
            self.notas = {}

    def definirNome(self, nome):
        self.nome = nome

    def addNota(self, disciplina, nota):
        self.notas[disciplina] = nota

    def calcularMedia(self):
        total = sum(self.notas.values())
        return total / len(self.notas)

# Exemplo de uso;
if __name__ == "__main__":
    #instanciando o Aluno:
    aluno1 = Aluno()
    aluno1.definirNome("João da Silva")

    #Adicionar Notas:
    aluno1.addNota("Matemática", 8.0)
    aluno1.addNota("Portugês", 7.0)
    aluno1.addNota("História", 9.0)

    #Obter a média de notas:
    print(f"Média de {aluno1.nome}: {aluno1.calcularMedia()}")

     #instanciando o Aluno:
    aluno2 = Aluno()
    aluno2.definirNome("Maria da Silva")

    #Adicionar Notas:
    aluno2.addNota("Matemática", 9.0)
    aluno2.addNota("Portugês", 8.0)
    aluno2.addNota("História", 10.0)

    #Obter a média de notas:
    print(f"Média de {aluno2.nome}: {aluno2.calcularMedia()}")
    

File: TecnicasProgram-II/Exemplos/Singleton/Singleton02.js
class Aluno{
    constructor(){
        if (!Aluno.instancia){
            this.nome = ""
            this.notas = {};
            Aluno.instancia = this;
        }
        return Aluno.instancia;
    }
    
    definirNome(nome){
        this.nome = nome;
    }

    addNotas(disciplina, notas){
        this.notas[disciplina] = notas;
    }

    calcularMedia(){
        const total = Object.values(this.notas).reduce((acc, nota) => acc + nota, 0);
        return total / Object.keys(this.notas).length;
    }
}

//Exemplo de uso:
const aluno1 = new Aluno();
aluno1.definirNome("João da Silva");
aluno1.addNotas("Matemática", 8.5);
aluno1.addNotas("Português", 7.0);
aluno1.addNotas("História", 9.2);
console.log(`Média de ${aluno1.nome}: ${aluno1.calcularMedia()}`);

const aluno2 = new Aluno();
aluno2.definirNome("Maria da Silva");
aluno2.addNotas("Matemática", 9.0);
aluno2.addNotas("Português", 5.0);
aluno2.addNotas("História", 8.0);
console.log(`Média de ${aluno2.nome}: ${aluno1.calcularMedia()}`);

console.log(aluno1 === aluno2);